typing/includecore.ml

```diff
@@ -378,6 +392,19 @@ let report_private_object_mismatch env ppf err =
   | Missing s -> pr "The implementation is missing the method %s" s
   | Types err -> report_type_inequality env ppf err
 
+let report_kind_mismatch first second ppf (kind1, kind2) =
+  let pr fmt = Format.fprintf ppf fmt in
+  let kind_to_string = function
+  | Kind_abstract -> "abstract"
+  | Kind_record -> "a record"
+  | Kind_variant -> "a variant"
+  | Kind_open -> "an extensible variant" in
+  pr "%s is %s, but %s is %s."
+    (String.capitalize_ascii second)
+    (kind_to_string kind2)
+    first
+    (kind_to_string kind1)
```

Comment:
Regarding the order: I switched `first` and `second` because I thought that "This variant or record ... This is ..., but the original is ..." was perhaps clearer because in both cases "this" comes first, e.g.:

(`second` then `first`, the current implementation)

```
Error: This variant or record definition does not match that of type
         ('a, [> `A ]) def
       This is a variant, but the original is a record.
```

Versus `first` then `second`

```
Error: This variant or record definition does not match that of type
         ('a, [> `A ]) def
       The original is a record, but this is a variant.
```

However in other contexts, it seems strange:

(`second` then `first`, the current implementation)

```
        The second is a record, but the first is a variant.
```

Versus `first` then `second`

```
        The first is a variant, but the second is a record.
```

---

I'm inclined to switch the order to `first` then `second` the get rid of "The second is ..., but the first is ...". Any thoughts on this?

Reply:
I've switched the order to `first` then `second`.

Reply:
I would suggest to keep the `first` then `second` order.
If we want to optimize the message order later, the right fix from my point of view is to compose the error message at the phrase level and avoid inserting words that accidentally fit in the grammatical context. 
This is a necessity for localization, but I think that this is a good guideline in general.